### ISSUE TITLE
Allow explicitly defining interfaces in GDScript

### ISSUE THREAD

**Issue by nebs** [godotengine/godot-proposals#4872]
### Describe the project you are working on

I'm working on a game where different entities can share similar behaviors, but this proposal is specific to GDScript, so it can apply to a wide range of projects.

### Describe the problem or limitation you are having in your project

Let's say we're making a simple shooter game. The player can shoot a `Bullet`. Bullets can hit a `Ship` and `Building`. There are also `Tree` entities that cannot be hit by bullets.

Here's some simple code (unrelated code omitted for simplicity).

**Ship.gd**
```gdscript
extends Node2D

class_name Ship

func die() -> void:
  print("Died")
```
**Building.gd**
```gdscript
extends Node2D

class_name Building

func die() -> void:
  print("Died")
```
**Tree.gd**
```gdscript
extends Node2D

class_name Tree

func die() -> void:
  print("Died")
```
**Bullet.gd**
```gdscript
extends Node2D

class_name Bullet

func on_body_entered(body: Node2D) -> void:
  if body.has_method("die"):
    body.die()
```

There are a few problems with this duck-typing approach:

1. The function "die" is referenced as a string which is error-prone
2. In `Bullet.gd` there is no code auto-completion or checking when calling `body.die()` so you need to implicitly know that that function exists.
3. The mere presence of a function named `die` is not enough to determine what kind of object you're dealing with. In this example, we don't want bullets to hit `Tree`, even though a tree might happen to have a function named `die` (the tree could die of natural causes for example).
4. If we wanted to make sure that the body we collided with implements more than just the `die` function we would have to do multiple `has_method` checks which can get ugly in code.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

My proposal is to introduce a language feature which allows us to explicitly define interfaces. This is a common feature in many modern languages and makes it easy to deal with complex object interactions without relying on inheritance.

See below for an implementation proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Let's take the same example code as above, and let's see what it could look like if GDScript had interface support.

**Killable.gd**
```gdscript
interface_name Killable

func die() -> void
```

**Ship.gd**
```gdscript
extends Node2D
implements Killable

class_name Ship

func die() -> void:
  print("Died")
```
**Building.gd**
```gdscript
extends Node2D
implements Killable

class_name Building

func die() -> void:
  print("Died")
```
**Tree.gd**
```gdscript
extends Node2D

class_name Tree

func die() -> void:
  print("Died")
```
**Bullet.gd**
```gdscript
class_name Bullet

func on_body_entered(body: Node2D) -> void:
  if body implements Killable:
    body.die() 
```
_Note: Because we're nested in an `implements` check we should be able to auto-complete "die" and also throw an error if "die" is not part of the Killable interface._

Alternatively, you can also reference the object via its interface by casting, for example:

**Bullet.gd**
```gdscript
class_name Bullet

func on_body_entered(body: Node2D) -> void:
  var killable: Killable = body as Killable
  if killable:
    killable.die()
```

This has a few advantages:

1. If you accidentally delete or rename `die()` from `Ship` or `Building`, the editor should show an error that you haven't fully implemented the interface. 
2. In `Bullet` we can now make our intentions explicit. The fact that there's a function called `die` is an implementation detail, but the true intent is we want to know if we're dealing with a `Killable` object. This way, classes like `Tree` which happen to have a `die` function won't be processed because they're not killable.
3. If the function `die` is renamed in the interface, then there would be errors showing all the places where it is being called and implemented. Previously the `has_method("die")` would not be flagged as an error because it's referencing the function as a string.
4. In `Bullet`, since we're now declared explicit intent that we want to deal with a `Killable` object, we should be able to auto-complete `Killable` functions, and also throw errors if we try to call a function that isn't defined in `Killable`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Technically it can be worked around with the current duck-typing approach. But I think it would make the code cleaner and safer to use explicit Interfaces. It's also easier and faster to write because we would have explicit checks and auto-completion. It's safer because it's more resilient to changes too (e.g. renaming a function wouldn't be flagged if the function is referenced as a string).

### Is there a reason why this should be core and not an add-on in the asset library?

Interfaces are built-in to other languages, I think this would make sense to be a core part of GDScript and not some external tool. It's akin to classes, functions, variables, etc. so I think it deserved first-class citizenship.

**Comment by nebs** [comment]
Another thing I forgot to mention, but with interfaces we can also be explicit about function arguments, for example:

```
func kill(killable: Killable) -> void:
  killable.die()
```

Whereas currently we'd have to do something like this:
```
func kill(node: Node2D) -> void:
  if node.has_method("die"):
    node.die()
```

**Comment by me2beats** [comment]
Can this be used in cases when Godot throws
cyclic dependency

**Comment by nsrosenqvist** [comment]
Yes please, in my own projects I have implemented a system for interfaces in GDScript for such cases where inheritance doesn't make sense, but I'd much rather have it be natively available

**Comment by sairam4123** [comment]
> Yes please, in my own projects I have implemented a system for interfaces in GDScript for such cases where inheritance doesn't make sense, but I'd much rather have it be natively available

It would be awesome if you could tell me how you did it, since I need this the most for my own projects.

**Comment by nsrosenqvist** [comment]
@sairam4123 It's not properly implemented at a language level, but it has worked well enough for my current project. I'll upload it and send you a link when I get home

**Comment by bitbrain** [comment]
This certainly would be absolutely helpful.

**Comment by Calinou** [comment]
@bitbrain Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by nsrosenqvist** [comment]
@sairam4123 @bitbrain sorry for the delay. I decided to clean it up a bit when I extracted the code from my utility library. You can now find the project as an addon here: https://github.com/nsrosenqvist/gdscript-interfaces/

**Comment by trommlbomml** [comment]
Would it make sense that the interface is just available before JIT kinda similiar to interfaces in typescript? This would have the benefit that interfaces have zero runtime overhead but fulfills this use case - with the trade off that at runtime you have no information about the interfaces any more. but you can argue that you have no real architectural improvement if you need to ask a class if it implements an interface and it may be easier at least as a step zero to get this feature.

**Comment by just-like-that** [comment]
I like your proposal.
I prefer to name it just interface and NOT interface_name coz the purpose is to define an interface (well with a specific name ofc).
2nd I'm opting for support of multiple interfaces of implementing classes.
Maybe like so: 
`class A implements Interface_A, Interface_B, ...
`

**Comment by nsrosenqvist** [comment]
@just-like-that I agree, otherwise they too get too bound up with inheritance and don't enable the flexibility sought after. Ideally should traits also be added at the same time as interfaces are implemented. See https://github.com/godotengine/godot-proposals/issues/758#

**Comment by nebs** [comment]
> I like your proposal. I prefer to name it just interface and NOT interface_name coz the purpose is to define an interface (well with a specific name ofc). 2nd I'm opting for support of multiple interfaces of implementing classes. Maybe like so: `class A implements Interface_A, Interface_B, ... `

So I agree with you in general that `interface` would read better than `interface_name`, but in my proposal I was trying to stay as close as possible to the current gdscript standards so as to minimize the friction for implementing this feature.

For example currently we set class names like this:
`class_name Foo`
So the closest analog for interfaces would be:
`interface_name Bar`

Similarly, we declare inheritance like this:
`extends Foo`
So the closest analog is:
`implements Bar`

I agree that the ability to implement multiple interfaces should be allowed. That was sort of implied but I guess I should have made that more clear in my original post.

I like your idea of having a new syntax like `class A implement IA, IB` etc but I feel like that would require a first unrelated change of adding the `class` keyword to gdscript, so I didn't want to add additional requirements for this particular issue.

In my original proposal, perhaps multiple interfaces can be implemented as so:
```
implements IA
implements IB
```

Which is a bit wordy, but I think it's okay as a first implementation because I can't imagine a class having to implement more than a handful of interfaces at a time. But perhaps we can add a comma system to this too.
```
implements IA, IB
```

Anyway, I suppose that's getting into the weeds a bit. I'd be happy with any of the implementations mentioned above. Whichever approach has the least friction is most likely to get implemented. I suppose we first need to get philosophical buy-in from the decision-makers.

**Comment by samdze** [comment]
What about the ability to also add property declarations in addition to methods as part of interfaces?

**Comment by nsrosenqvist** [comment]
> What about the ability to also add property declarations in addition to methods as part of interfaces? Signal declarations maybe? That would be nice.

I think signals makes sense, but not properties. They are usually dependent on how one chooses to implement an interface (eg. a data property that the methods operate on, or if another implementation instead fetches data from a remote source and don't do operate on data locally).

**Comment by samdze** [comment]
> I think signals makes sense, but not properties. They are usually dependent on how one chooses to implement an interface (eg. a data property that the methods operate on, or if another implementation instead fetches data from a remote source and don't do operate on data locally).

Still possible, as long as properties implemented as part of an interface are forced to have a getter and (optional?) a setter.

**Comment by DasGandlaf** [comment]
YES!! Interfaces. That's the biggest reason I am using c# at the moment.
I think GDScript interfaces should work exactly the way they do in c#. It's simple for the user, yet extremely helpful.
This way, my team and I could finally start automatically testing our GDScript code, while still having the benefits of static-typed GDScript.

**Comment by just-like-that** [comment]
> So I agree with you in general that `interface` would read better than `interface_name`, but in my proposal I was trying to stay as close as possible to the current gdscript standards so as to minimize the friction for implementing this feature.

From my understanding it's called class_name coz it's the name for the whole class file.
A class file contains consts, functions (also static), variables and inner classes.
A class_name has to be unique thru the whole project.

Inner classes are named "**class**" only.
So interface_name would indicate it's also in its own file and they have to be unique thru out the prj.
Was this your initial idea? Binding of interfaces in its own interface file?

I see an interface more as a part of a specific class file.
As such a class file can contain more than one interface declaration.
The class file acts as a context aka name space to that interface.
This has the benefit that interfaces with the same name i.e. from plugins don't interfere with your own interface declarations.

By pre-loading the class file via
`const <Example_Class> := preload("<example-class-path>")`
it's possible to reference each interface individually without any interference.

`Example_Class.Interface_A`

**Comment by dalexeev** [comment]
> By pre-loading the class file via `const <Example_Class> := preload("<example-class-path>")` it's possible to reference each interface individually without any interference like so
> 
> `Example_Class.Interface_A`

If you need to make the interface global, see my suggestion #4740.

**Comment by willnationsdev** [comment]
If this were to be a thing in GDScript, I would suggest just having an `@interface` annotation which you could then apply above the `class` or `class_name` keywords to indicate that an inner class or the top-level class is an interface. That will then avoid confusion related to differentiating between `interface` and `interface_name`, remove the need for `interface_name` altogether, and avoid populating the global namespace with more keywords. You could then similarly define an `@global` annotation to generate global variables for inner interfaces throughout the GDScript language similar to what #4740 proposes.

Regardless of whether an interface system is added though, it'd be important to identify exactly how it would work. The most prevalent forms of interfaces I have observed are those in TypeScript and in C#, but the way they work in each of those languages is vastly different, so people should iron out what they actually want GDScript's interfaces to *be* first. I'll refer to these as the "Strict"/C# system vs. the "Loose"/TypeScript system.

## How is an interface evaluated against a type?

"Strict": Is an interface a specific type with specific methods and a given script must explicitly implement the interface in order for it to be recognized as an implementation of the interface?

```gdscript
@interface
class IKillable:
    ## returns points awarded when killed. Prints `p_msg` afterward.
    func die(p_msg: String) -> int:
        pass
# parse error if no `pass`. Could also be on the next line.
# Must have name `die`.
# Must have 1 argument of type `String`. The name does not need to be `p_msg`.
# Must have `int` return type.
# GDScript parser would need to ignore improper return types on methods with static type hints.

class Ship implements IKillable:
    pass
# will not compile until a `die` method matching all the above conditions is defined.
```

OR

"Loose": Is an interface just a way of checking multiple `has_method` calls implicitly?

```gdscript
@interface
class IKillable:
    func die(p_msg: String) -> int:
        pass
# Exact same as the Strict case

class Ship:
    func die(p_msg):
        pass
# This satisfies an `if ship implements IKillable` check b/c `has_method(...)` returns true for all subset of methods in IKillable.

OR

"Hybrid": is it still any implicit union of types, but where it actually tracks and enforces static type checks, i.e. is it a union of methods with specific names, parameter count, types for each parameter, and a specific return type?

```gdscript
```gdscript
# Exact same `IKillable` as the Loose case

class Ship:
    func die(p_msg: String) -> int:
        return 3
# This satisfies an `if ship implements IKillable` check b/c all signatures of all methods in `IKillable` match exactly
# EVEN THOUGH the class never explicitly says it matches `IKillable` (so no static checking of interface conformity).

```

## How does the interface handle edge cases related to method compatibility?

"Strict": what about situations where a class implements two interfaces that share method names, but different arguments?

```gdscript
@interface
class IBag:
    func configure(p_slot_count: int = 0) -> void: pass

@interface
class IMail:
    func configure(p_owner: StringName, p_content: String) -> void: pass

class PostalDelivery implements IBag, IMail:
    # <implements methods exactly, but triggers parse error>
    pass
# GDScript does not support method overloading.
# Do we just forbid these combinations? That'd be a tough sell
# as people start relying on third-party libraries that all define global script classes.

# For "Strict" only, would we need to resort to having interface-specific annotations on methods to flag them?
# You'd need to make these hidden methods that aren't publicly accessible, likely with a generated name
# derived from the interface it belongs to (that way it doesn't actually occupy the same slot in the method hashmap
# used by the GDScript's C++ engine code).
    @interface(IBag)
    func configure(p_slot_count: int = 0) -> void: pass
    @interface(IMail)
    func configure(p_owner: StringName, p_content: String) -> void: pass
    # ^ This duplication of method name definitions would need to NOT trigger a parse error.
```

"Loose": what about methods with an extra but *optional* variable that has a default value - would it still be an implementation?

```gdscript
@interface
class IKillable:
    func die(): pass

class Ship
    func die(p_msg := "") -> void: # does
        print(p_msg if p_msg != null else "ship died")
    # does this method count? It's POSSIBLE to call w/o args, so it WOULD satisfy duck-typing,
    # but it doesn't explicitly match the interface definition.
```

"Hybrid" behaves the same as "Loose".

## Do interfaces support unions and/or inheritance?

"Strict":

```gdscript
@interface
class_name IShip
extends IKillable, IRevivable
# has methods `die()` and `revive()`, but no content in script.
```

"Loose": Could you then later define an interface that is a union of other interfaces?

```gdscript
@interface
class_name IShip
extends IKillable & IRevivable
# has methods `die()` and `revive()`, but no content in script.
# Because interfaces are mere combinations of methods, you can use bitwise operators on them (potentially).
```

"Hybrid" behaves the same as "Loose" in this regard too.

Also, since static typing is optional, what does static type checking on interfaces *mean* when a method on an interface *does not use static type hints?* Do you just reduce the number of required conditions an implementing class has to meet in order to conform to the interface? Or do you consider it a compiler error and demand that all methods on an interface use type hints everywhere?

---

I think it would be possible to go with any of the 3 versions. "Loose" would obviously be the most straightforward and simple to implement, but I doubt it would satisfy people who are looking for full static type checks to be honored by interface declarations.

Between the "Hybrid" and "Strict" versions, I think "Hybrid" more closely mimics GDScript's "scripting" context, but at the same time feel conflicted since "Strict" more closely honors GDScript's "pythonic C++" style. Ultimately though, GDScript is a scripting language and is meant to be easy to use. Regardless of which approach gets general consensus, I believe that interfaces should not be introduced in such a way that they only work when users elect to use static typing. They must also work with a complete lack of static types. Given that "duck-typing" is already the style used when a lack of static types is present, I would opt to take a "Hybrid" approach as it is the one that is most compatible with dynamic GDScript, and it makes GDScript behave like other scripting languages with type hints.

**Comment by just-like-that** [comment]
What about this temporary solution for now:

Interface definition:

```
# interface_a.gd
extends Reference

## Returns an random integer between 0 and max value (inclusive).
func irnd(max_val: int) -> int:
	assert(false, "InterfaceNotCallableException.")
	# TODO: needs implementation in implementing class
	return -1
###
```

Implementing class:

```
# my_class.gd
extends Reference

const Interface_A := preload("res://scene/interfaces/interface_a.gd")

# implements Interface_A
func irnd(max_val: int) -> int:
	return randi() % (max_val + 1)
###

func is_class(clazz: String) -> bool:
	# check against own class and all its (directly) implementing interfaces
	return clazz in ["MyClass", Interface_A.get_path()] || .is_class(clazz)
###
```

Usage example:

```
# interface.gd
extends Node

const MyClass := preload("res://scene/interfaces/my_class.gd")

const Interface_A := preload("res://scene/interfaces/interface_a.gd")


func _ready() -> void:
	var implementing_class := MyClass.new()
	prints("class check:", implementing_class.is_class("MyClass"), true)
	prints("class check:", implementing_class.is_class("NotMyClass"), false)
	prints("interface check:", implementing_class.is_class(Interface_A.get_path()), true)
	prints("interface check:", implementing_class.is_class("res://scene/interfaces/interface_b.gd"), false)

	_do_action_on(implementing_class)
###


func _do_action_on(instance: Object) -> void:
	if instance.is_class(Interface_A.get_path()) :
		prints("random int value=", instance.irnd(7))
###
```

**Comment by DasGandlaf** [comment]
@willnationsdev I would prefer the approach of every single one being strict, because that just makes the most sense, when talking about interfaces, and what they are actually supposed to do. They are supposed to be strict. Making them only half-check makes no sense I think. Because that is the whole point of them.
Also, in one of your examples, you set a default value of a parameter on an interface. I don't think that should be done in the interface, but rather in the implementation. The interface should just say, "You have the add function, with returns a number. It has two parameters, which are both numbers. What default value they hold, is not up to me, since you are the implementation. But you cannot add more parameters, even if they are just optional, since I told you to make a method with only two parameters, that's it. Once you add another parameter, I will throw an error, since you did not implement what I told you to implement.".
Interfaces are like contracts, and the implementation is obliged to follow that contract. Whether a new optional condition gets added or not, is not up to the implementation to decide, but it is written in the contract if that optional condition is there or not.

> Regardless of which approach gets general consensus, I believe that interfaces should not be introduced in such a way that they only work when users elect to use static typing. They must also work with a complete lack of static types. Given that "duck-typing" is already the style used when a lack of static types is present, I would opt to take a "Hybrid" approach as it is the one that is most compatible with dynamic GDScript, and it makes GDScript behave like other scripting languages with type hints.

Interfaces and duck typing contradict each other. Interfaces are there to check whether this class has the methods from the interface it's implementing. With duck typing, you just try to find out what it has and call it. It only throws an error, once it hits that line and finds out, `Oh wait, that object doesn't even have this method`. I don't see how and why they should be "merged" in this type of way. If you add interfaces, don't do it halfway. It would not satisfy the duck typers, since they won't use interfaces, nor it would satisfy the static typers, which would get half of an interface.

**Comment by DasGandlaf** [comment]
I just wrote a 400000 paragraph comment responding to the comments here, then I realized, I just want interfaces lol.
Are there any news on this?

**Comment by Calinou** [comment]
> Are there any news on this?

Godot 4.0 is in feature freeze. To allow the language to stabilize, no new GDScript features will be added until after 4.0 is released.

**Comment by matiturock** [comment]
So, for now the interfaces are not going to be implemented. Is it to promote composition over inheritance?

**Comment by Ejsstiil** [comment]
Is there any news on this?

**Comment by Calinou** [comment]
> Is there any news on this?

To my knowledge, nobody is currently working on implementing this.

**Comment by ryanabx** [comment]
> > Is there any news on this?
> 
> To my knowledge, nobody is currently working on implementing this. 4.2 is in feature freeze, so any new features will be for 4.3 at the earliest.

I need this for my project, and I've already implemented abstract methods for 4.3 or beyond, in https://github.com/godotengine/godot/pull/82987 , so it's time to take a stab at this

**Comment by Kalto-Mate** [comment]
What I don't understand is this kind of features being outright missing. Makes me think all those Godot preachers that finally got their wish after Unity did a silly NEVER actually sat down and made something remotely complex and robust

**Comment by Calinou** [comment]
@Kalto-Mate I have to remind you that we have a [Code of Conduct](https://godotengine.org/code-of-conduct). Please stay constructive.

**Comment by LEEROY-3** [comment]
It's a shame this feature has not been implemented initially for whatever reason. This feature missing from gdscript means C# interfaces can't be utilized to it's max potential either. See #7971, for example: A potentially powerful feature that isn't particularly hard to implement but will not be implemented as we must respect gdscript first, which does not have interfaces.

At this stage of gdscript, I imagine, it will be incredibly hard to implement such a feature as the language has been built around _not_ having interfaces, which might explain why nobody is particularly eager to pick up this issue. Just a saddening situation all in all.

**Comment by ryanabx** [comment]
> At this stage of gdscript, I imagine, it will be incredibly hard to implement such a feature as the language has been built around _not_ having interfaces, which might explain why nobody is too eager to pick up this issue. Just a saddening situation all in all.

It's actually not that hard to implement, it just requires a few building blocks first.

https://github.com/godotengine/godot/pull/67777

A PR which implements an `@abstract` annotation for GDScript. This PR only covers classes, not methods. Potentially slated for 4.3


https://github.com/godotengine/godot/pull/82987

A PR built on top of the other one which implements abstract methods through the same keyword. Also potentially slated for 4.3.

With these two PRs, it should be easy to make interfaces, just forbid any other members except abstract functions in an interface class and create an `implements` keyword.

I actually have a branch on my godot repo fork that has most of the implementation finished for interfaces, all that needs to be done is resolving classes that are implemented by another class.

https://github.com/ryanabx/godot/tree/gdscript/interfaces

The main reason I stopped working on it is because another godot engine contributor @vnen is going to eventually work on Trait support for GDScript, which would be a superset of the features that interfaces provide.

Believe me, I'm also anxiously waiting for interface support in GDScript, but it seems we will have to wait a little bit longer for something like it to be implemented in Godot.

**Comment by Calinou** [comment]
> Has there been any progress towards this at all since the last post? I am really interested to get this feature.

See the above comment:

> The main reason I stopped working on it is because another godot engine contributor @vnen is going to eventually work on Trait support for GDScript, which would be a superset of the features that interfaces provide.

**Comment by cyda89** [comment]
+1 here. As a scala developer i cant wait to get traits in gdscript!

**Comment by Calinou** [comment]
@cyda89 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by hidemat** [comment]
> The main reason I stopped working on it is because another godot engine contributor @vnen is going to eventually work on Trait support for GDScript, which would be a superset of the features that interfaces provide.

This is my first time hearing about traits, but I looked up how they are used in php:
https://www.youtube.com/watch?v=x9bj30cWolA

If this is the implementation the gdscript is going for, wouldn't both Traits and Interfaces be desirable? It's seems to me the are NOT interchangeable and have different use cases.

**Comment by interacsion** [comment]
> > The main reason I stopped working on it is because another godot engine contributor [@vnen](https://github.com/vnen) is going to eventually work on Trait support for GDScript, which would be a superset of the features that interfaces provide.
> 
> This is my first time hearing about traits, but I looked up how they are used in php: https://www.youtube.com/watch?v=x9bj30cWolA
> 
> If this is the implementation the gdscript is going for, wouldn't both Traits and Interfaces be desirable? It seems to me they are NOT interchangeable and have different use cases.

It appears that PHP makes a distinction between Interfaces which can only contain method signatures, and Traits which must provide concrete implementations. As far as I am aware Java, C#, and Rust don't make such a distinction, and let you have both in the same type.

**Comment by Diomeh** [comment]
Has there been any progress on either traits or interfaces? We're at Godot v4.4 already and the only solutions I've been able to find are hacks around GDScript and crude implementations with a lot of limitations.

**Comment by RudyFisher7** [comment]
Question for the OP. Would a component system satisfy this use case? I mean, component system like seen in Unity/Flax engines where you have Entity/Actor and attach script components to them, than can do a query on Actor.has_component(&"Killeable")?

**Comment by Rito13** [comment]
> Question for the OP. Would a component system satisfy this use case? I mean, component system like seen in Unity/Flax engines where you have Entity/Actor and attach script components to them, than can do a query on Actor.has_component(&"Killeable")? So essentially this would just allow multiple scripts in any Node that could be queried.

Interfaces allow to use different classes in the same way without knowing which of them is currently being used. For example array could be implemented in 3 different ways: as a list, as a vector and as a binary tree. Each of them has advantages in different circumstances but for some class like parent which holds array of children it does not matter how array is implemented but what methods and properties it has.

**Comment by RudyFisher7** [comment]
> > Question for the OP. Would a component system satisfy this use case? I mean, component system like seen in Unity/Flax engines where you have Entity/Actor and attach script components to them, than can do a query on Actor.has_component(&"Killeable")? So essentially this would just allow multiple scripts in any Node that could be queried.
> 
> Interfaces allow to use different classes in the same way without knowing which of them is currently being used. For example array could be implemented in 3 different ways: as a list, as a vector and as a binary tree. Each of them has advantages in different circumstances but for some class like parent which holds array of children it does not matter how array is implemented but what methods and properties it has. 

Yes. This can be done with components. Essentially, you would attach an ArrayComponent. This defines the public facing API (the functions other objects csn use). Internally it uses an Array extends Resource. You can implement ArrayTree extends Array, List extends Array and choose with implementation you want ArrayComponent to have via inspector.

**Comment by Rito13** [comment]
> It's actually not that hard to implement, it just requires a few building blocks first.

Basically with use of duck typing only syntax checks are needed which can be performed in the editor before runtime. 

> Believe me, I'm also anxiously waiting for interface support in GDScript, but it seems we will have to wait a little bit longer for something like it to be implemented in Godot.

I got sick of waiting for traits and have implemented this feature in GDScript it self as a [Godot plugin/asset](https://github.com/Rito13/GDScript-Interfaces-addon). It will not boost performance because it converts static typing into runtime type detection. Despite that it simplify scripts and allow for in editor code verification also it shows how powerfull GDScript is.

**Comment by flamenco687** [comment]
Just a heads up that, finally, a trait (you could really call it interface) feature is actively [being discussed](https://github.com/godotengine/godot-proposals/discussions/12567) for 4.6. However, note that the current [PR](https://github.com/godotengine/godot/pull/107227) does [not address type features](https://github.com/godotengine/godot-proposals/discussions/12567#discussioncomment-13442537), which is kind of the point, right? In any case, it has already been stated how traits and interfaces differ merely in names, so maybe this issue should now link to this proposal.

**Comment by aduernberger** [comment]
The current PR is a cut down version of https://github.com/godotengine/godot/pull/97657 which does implement `is`. The original PR was too big so a new minimal PR was created. Features will probably be added on top if/when the new PR gets accepted.

**Comment by eobet** [comment]
I might not be able to grasp everything which has been said here, but most of the examples I've seen still checks for interface types before engaging with the interface. Wouldn't it be nice to be able to skip that step?

Has there been any thought put into broadcasting messages via an interface without knowing or caring if it is received or not? Sort of like an interface signal that doesn't need a connection (or a "signal bus" implementation)?

**Comment by diomeh** [comment]
> I might not be able to grasp everything which has been said here, but most of the examples I've seen still checks for interface types before engaging with the interface. Wouldn't it be nice to be able to skip that step?
> 
> Has there been any thought put into broadcasting messages via an interface without knowing or caring if it is received or not? Sort of like an interface signal that doesn't need a connection (or a "signal bus" implementation)?

I don't think an interface and a signal bus are interchangeable as they fulfill different roles. 

A signal bus allows different systems to communicate with each other without the need for coupling. An interface on the other hand allows classes to share common functionality without the need for code repetition.

Currently implementing a signal bus is as simple as defining all shared signals in a singleton script and then connecting / emitting signals within a script, as needed.

**Comment by ArcanumDieM** [comment]
I tried the new abstract class in Godot 4.5: https://godotengine.org/releases/4.5/#abstract-classes-and-methods
And it is very like an interface, for a basic usage can already be a good solution, i think

**Comment by flamenco687** [comment]
@ArcanumDieM I think you are misunderstanding the use case of interfaces 😛. Interfaces are intended to provide common functionality to multiple different classes that do not necessarily inherit from the same parent class.

Abstract classes are still constrained to their definitions; animals: things that move and are alive; vehicles: things that move but aren't alive. However, interfaces are more general and may act, specially, as components. An example would be a "Health" interface for both animals and vehicles. Cars and geckos have nothing to do and are not compatible class-wise, but they can share health functionality.

**Comment by rshouker** [comment]
Yes, please. GDscript has introduced many strict typing options lately, this one is sorely missing.
This should be integrated with GDExtension and the .NET support for interoperability, so you could introduce the interface in C# or C++ (with proper registration) and implement it on GDScript.

**Comment by dylanmartinezs** [comment]
Is there any news on this? Can't believe something so basic but crucial isn't implemented yet

**Comment by Naros** [comment]
It isn't as basic a concept as you might think.

Creating and referring to object types by their interfaces requires some sort of ClassDB or GDType registration mechanism for interfaces. That itself would extend to the entire Godot ecosystem, not just GDScript, providing a solution for the engine, Scripting/ScriptExtension (GDScript, CSharp), and GDExtension. It should be done in a way that checks all these boxes, because it's not a feature that is only useful to GDScript.

**Comment by clarence112** [comment]
> Is there any news on this? Can't believe something so basic but crucial isn't implemented yet

These are the current PRs related to this feature:  
https://github.com/godotengine/godot/pull/107227  
https://github.com/godotengine/godot/pull/107608  
https://github.com/godotengine/godot/pull/112933  
(Currently seems to be blocked by the unified type system)

**Comment by solodeveloping** [comment]
I'm not sure if this is the right place, or if it has been mentioned elsewhere before, but there are tons of issues on these topics, so 🤷‍♂️

The idea is: allow the @abstract class system to act more like an interface

In typescript, you can do this:

```ts

interface titi {
  name: string
}

interface tutu {
  name: string
  age: string
}

function dosomething(titi: titi) {
  // type autocompletion and such
}

var yo: tutu = {
  name: "Toto",
  age: "30"
}

// no errors
dosomething(yo)

```

In Godot, I can't assign a variable having the same signature as an abstract class in the editor, and in the code, if I try, I'll have an error.

This is useful when you are creating code in addons, you don't necessarily care about how the end user will implement his own class, and if there is 2 addon having the same expectation, you don't have to create a third addon to share the interfaces.

Interface, abstract class etc It's a fuzzy concept, this is how duck typing is implemented in typescript.

I'm wondering if the abstract class system could be modified to allow for this.
